Replace react-sortable-hoc with @dnd-kit/sortable

[react-sortable-hoc](https://www.npmjs.com/package/react-sortable-hoc) is based on deprecated features in React, and will no longer work once `findDOMNode` is removed.  Given that this library is only used in SortablePuzzleList, it should not take significant development effort to migrate to [dnd-kit/sortable](https://www.npmjs.com/package/@dnd-kit/sortable).

From the [react-sortable-hoc README](https://github.com/clauderic/react-sortable-hoc#%EF%B8%8F-important-notice-for-new-consumers):

>### ⚠️ Important notice for new consumers
>
>New consumers are strongly encouraged to check out [@dnd-kit](https://github.com/clauderic/dnd-kit) before adopting this library in a new project. It supports all of the features of react-sortable-hoc, but with a more modern and extensible architecture.
>
>This library will continue to receive critical updates and bug fixes for the foreseeable future, but there are no new features planned. In future versions of React, the [findDOMNode](https://reactjs.org/docs/react-dom.html#finddomnode) method will be deprecated. This method is a critical piece of the architecture of react-sortable-hoc, and it's unlikely that the library could continue to exist with its current architecture once this method is deprecated. In light of this, all development efforts have been redirected towards [@dnd-kit](https://github.com/clauderic/dnd-kit).

### Motivation

This is mostly just on my radar because every time I run `tsc` I get this error:

```
$ npx tsc
node_modules/react-sortable-hoc/types/index.d.ts:108:11 - error TS2694: Namespace 'React' has no exported member 'SFC'.

108   | React.SFC<P>
              ~~~


Found 1 error in node_modules/react-sortable-hoc/types/index.d.ts:108
```

### See also

- https://docs.dndkit.com/presets/sortable

--- a/package.json
+++ b/package.json
@@ -75,9 +75,11 @@
         "webpack-bundle-analyzer": "^4.5.0",
         "webpack-cli": "^4.9.2",
         "xgettext-js": "^3.0.0",
-        "yarn": "^1.22.18"
+        "yarn": "^1.22.19"
     },
     "dependencies": {
+        "@dnd-kit/core": "^6.0.3",
+        "@dnd-kit/sortable": "^7.0.0",
         "@sentry/browser": "^6.16.1",
         "@types/d3": "^5.7.2",
         "@types/howler": "^2.2.2",
@@ -115,7 +117,6 @@
         "react-resize-detector": "^7.1.1",
         "react-router-dom": "^6.3.0",
         "react-select": "^5.0.1",
-        "react-sortable-hoc": "^2.0.0",
         "react-split": "^2.0.9",
         "react-switch": "^5.0.1",
         "react-table": "^6.10.0",

--- a/src/views/PuzzleCollection/SortablePuzzleList.tsx
+++ b/src/views/PuzzleCollection/SortablePuzzleList.tsx
@@ -21,8 +21,21 @@ import { errorAlerter, navigateTo } from "misc";
 import { get, put } from "requests";
 import { longRankString } from "rank_utils";
 import { MiniGoban } from "MiniGoban";
-import { SortableContainer, SortableElement } from "react-sortable-hoc";
-import arrayMove from "array-move";
+import {
+    DndContext,
+    DragEndEvent,
+    useSensors,
+    MouseSensor,
+    useSensor,
+    TouchSensor,
+} from "@dnd-kit/core";
+import { CSS } from "@dnd-kit/utilities";
+import {
+    arrayMove,
+    SortableContext,
+    useSortable,
+    verticalListSortingStrategy,
+} from "@dnd-kit/sortable";
 import { IdType } from "src/lib/types";
 
 interface PuzzleEntryInterface {
@@ -66,71 +79,112 @@ export class SortablePuzzleList extends React.Component<
             .catch(errorAlerter);
     }
 
-    onSortEnd = ({ oldIndex, newIndex }) => {
-        if (oldIndex === newIndex) {
+    handleDragEnd = (event: DragEndEvent) => {
+        const { active, over } = event;
+
+        if (over == null) {
             return;
         }
+
+        if (active.id === over.id) {
+            return;
+        }
+
+        const oldIndex = this.state.entries.findIndex((entry) => active.id === entry.id);
+        const newIndex = this.state.entries.findIndex((entry) => over.id === entry.id);
 
         const after = oldIndex < newIndex ? newIndex : newIndex - 1;
 
         put(`puzzles/${this.state.entries[oldIndex].id}/order`, {
             after: newIndex ? this.state.entries[after].id : 0,
-        })
-            .then((asdf) => {
-                console.log(asdf);
-            })
-            .catch(errorAlerter);
+        }).catch(errorAlerter);
 
-        this.setState(({ entries }) => ({
-            entries: arrayMove(entries, oldIndex, newIndex),
-        }));
+        this.setState(({ entries }) => {
+            return { entries: arrayMove(entries, oldIndex, newIndex) };
+        });
     };
 
     render() {
         return (
-            <SortablePuzzleListContainer entries={this.state.entries} onSortEnd={this.onSortEnd} />
+            <SortablePuzzleListContainer
+                entries={this.state.entries}
+                onDragEnd={this.handleDragEnd}
+            />
         );
     }
 }
 
-const PuzzleEntry: any = SortableElement(({ puzzle }: { puzzle: PuzzleEntryInterface }) => (
-    <li className="SortablePuzzleListEntry">
-        <span className="minigoban">
-            <MiniGoban
-                noLink
-                id={null}
-                json={puzzle.puzzle}
-                displayWidth={64}
-                white={null}
-                black={null}
-            />
-        </span>
-        <span className="name">{puzzle.name}</span>
-        <span className="difficulty">{longRankString(puzzle.rank)}</span>
-        <button
-            className="edit"
-            onClick={(ev) => {
-                ev.stopPropagation();
-                ev.preventDefault();
-                navigateTo(`/puzzle/${puzzle.id}`, ev);
-            }}
-            onAuxClick={(ev) => {
-                ev.stopPropagation();
-                ev.preventDefault();
-                navigateTo(`/puzzle/${puzzle.id}`, ev);
-            }}
-        >
-            {_("Edit")}
-        </button>
-    </li>
-));
+function PuzzleEntry({ puzzle }: { puzzle: PuzzleEntryInterface }) {
+    const { attributes, listeners, setNodeRef, transform, transition } = useSortable({
+        id: puzzle.id,
+    });
 
-const SortablePuzzleListContainer: any = SortableContainer(
-    ({ entries }: { entries: Array<PuzzleEntryInterface> }) => (
-        <ul className="SortablePuzzleList">
-            {entries.map((entry, index) => (
-                <PuzzleEntry key={entry.id} index={index} puzzle={entry} />
-            ))}
-        </ul>
-    ),
-);
+    const style = {
+        transform: CSS.Transform.toString(transform),
+        transition,
+    };
+
+    return (
+        <li
+            className="SortablePuzzleListEntry"
+            ref={setNodeRef}
+            {...attributes}
+            {...listeners}
+            style={style}
+        >
+            <span className="minigoban">
+                <MiniGoban
+                    noLink
+                    id={null}
+                    json={puzzle.puzzle}
+                    displayWidth={64}
+                    white={null}
+                    black={null}
+                />
+            </span>
+            <span className="name">{puzzle.name}</span>
+            <span className="difficulty">{longRankString(puzzle.rank)}</span>
+            <button
+                className="edit"
+                onClick={(ev) => {
+                    ev.stopPropagation();
+                    ev.preventDefault();
+                    navigateTo(`/puzzle/${puzzle.id}`, ev);
+                }}
+                onAuxClick={(ev) => {
+                    ev.stopPropagation();
+                    ev.preventDefault();
+                    navigateTo(`/puzzle/${puzzle.id}`, ev);
+                }}
+            >
+                {_("Edit")}
+            </button>
+        </li>
+    );
+}
+
+function SortablePuzzleListContainer({
+    entries,
+    onDragEnd,
+}: {
+    entries: Array<PuzzleEntryInterface>;
+    onDragEnd: (event: DragEndEvent) => void;
+}) {
+    const sensors = useSensors(
+        // Without this activation constraint, the "Edit" button doesn't work
+        // because dnd swallows click events.
+        useSensor(MouseSensor, { activationConstraint: { distance: 5 } }),
+        useSensor(TouchSensor),
+    );
+    return (
+        <DndContext onDragEnd={onDragEnd} sensors={sensors}>
+            <SortableContext items={entries} strategy={verticalListSortingStrategy}>
+                <ul className="SortablePuzzleList">
+                    {entries.map((entry) => (
+                        <PuzzleEntry key={entry.id} puzzle={entry} />
+                    ))}
+                </ul>
+            </SortableContext>
+        </DndContext>
+    );
+}

